```
[...]/src/storage/innobase/include/fil0fil.h:1495:16: error: invalid operand for inline asm constraint 'i'
  __asm__ goto("lock btsl $31, %0\t\njnc %l1" : : "m" (n_pending)
               ^
```

--- a/storage/innobase/include/fil0fil.h
+++ b/storage/innobase/include/fil0fil.h
@@ -1490,7 +1490,7 @@
 #if (defined __clang_major__ && __clang_major__ < 10) || defined __APPLE_CC__
   /* Only clang-10 introduced support for asm goto */
   return n_pending.fetch_or(STOPPING, std::memory_order_relaxed) & STOPPING;
-#elif defined __GNUC__ && (defined __i386__ || defined __x86_64__)
+#elif defined __GNUC__ && ((defined __i386__ && !defined __ANDROID__) || defined __x86_64__)
   static_assert(STOPPING == 1U << 31, "compatibility");
   __asm__ goto("lock btsl $31, %0\t\njnc %l1" : : "m" (n_pending)
                : "cc", "memory" : not_stopped);
--- a/storage/innobase/sync/srw_lock.cc
+++ b/storage/innobase/sync/srw_lock.cc
@@ -271,7 +271,7 @@
 /* Only clang-10 introduced support for asm goto */
 #elif defined __APPLE__
 /* At least some versions of Apple Xcode do not support asm goto */
-#elif defined __GNUC__ && (defined __i386__ || defined __x86_64__)
+#elif defined __GNUC__ && ((defined __i386__ && !defined __ANDROID__) || defined __x86_64__)
 # define IF_FETCH_OR_GOTO(mem, bit, label)				\
   __asm__ goto("lock btsl $" #bit ", %0\n\t"				\
                "jc %l1" : : "m" (mem) : "cc", "memory" : label);
